File: YukariLauncher/src/main/java/net/kdt/pojavlaunch/progresskeeper/ProgressKeeper.java
package net.kdt.pojavlaunch.progresskeeper;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * 進行状況を管理し、リスナーに通知するクラス。
 * 複数のタスクの進行状況を追跡し、タスク数の変更をリスナーに通知します。
 */
public class ProgressKeeper {
    private static final HashMap<String, List<ProgressListener>> sProgressListeners = new HashMap<>();
    private static final HashMap<String, ProgressState> sProgressStates = new HashMap<>();
    private static final List<TaskCountListener> sTaskCountListeners = new ArrayList<>();

    /**
     * 進行状況を送信します。進行状況が-1、residが-1の場合は終了を示します。
     */
    public static synchronized void submitProgress(String progressRecord, int progress, int resid, Object... va) {
        ProgressState progressState = sProgressStates.get(progressRecord);
        boolean shouldCallStarted = progressState == null;
        boolean shouldCallEnded = resid == -1 && progress == -1;
        if(shouldCallEnded) {
            shouldCallStarted = false;
            sProgressStates.remove(progressRecord);
            updateTaskCount();
        }else if(shouldCallStarted){
            sProgressStates.put(progressRecord, (progressState = new ProgressState()));
            updateTaskCount();
        }
        if(progressState != null) {
            progressState.progress = progress;
            progressState.resid = resid;
            progressState.varArg = va;
        }

        List<ProgressListener> progressListeners = sProgressListeners.get(progressRecord);
        if(progressListeners != null)
            for(ProgressListener listener : progressListeners) {
                    if(shouldCallStarted) listener.onProgressStarted();
                    else if(shouldCallEnded) listener.onProgressEnded();
                    else listener.onProgressUpdated(progress, resid, va);
            }
    }

    /**
     * タスク数を更新し、リスナーに通知します。
     */
    private static synchronized void updateTaskCount() {
        int count = sProgressStates.size();
        for(TaskCountListener listener : sTaskCountListeners) {
            listener.onUpdateTaskCount(count);
        }
    }

    /**
     * 特定の進行状況レコードにリスナーを追加します。
     */
    public static synchronized void addListener(String progressRecord, ProgressListener listener) {
        ProgressState state = sProgressStates.get(progressRecord);
        if(state != null && (state.resid != -1 || state.progress != -1)) {
            listener.onProgressStarted();
            listener.onProgressUpdated(state.progress, state.resid, state.varArg);
        }else{
            listener.onProgressEnded();
        }
        List<ProgressListener> listenerWeakReferenceList = sProgressListeners.computeIfAbsent(progressRecord, k -> new ArrayList<>());
        listenerWeakReferenceList.add(listener);
    }

    /**
     * 特定の進行状況レコードからリスナーを削除します。
     */
    public static synchronized void removeListener(String progressRecord, ProgressListener listener) {
        List<ProgressListener> listenerWeakReferenceList = sProgressListeners.get(progressRecord);
        if(listenerWeakReferenceList != null) listenerWeakReferenceList.remove(listener);
    }

    /**
     * タスク数変更リスナーを追加します。
     */
    public static synchronized void addTaskCountListener(TaskCountListener listener) {
        listener.onUpdateTaskCount(sProgressStates.size());
        if(!sTaskCountListeners.contains(listener)) sTaskCountListeners.add(listener);
    }

    /**
     * タスク数変更リスナーを追加します（初回更新の制御付き）。
     */
    public static synchronized void addTaskCountListener(TaskCountListener listener, boolean runUpdate) {
        if(runUpdate) listener.onUpdateTaskCount(sProgressStates.size());
        if(!sTaskCountListeners.contains(listener)) sTaskCountListeners.add(listener);
    }

    /**
     * タスク数変更リスナーを削除します。
     */
    public static synchronized void removeTaskCountListener(TaskCountListener listener) {
        sTaskCountListeners.remove(listener);
    }

    /**
     * @return 現在のタスクセット内に指定されたキーが存在するかどうか
     */
    public static boolean containsProgress(String progressKey) {
        return sProgressStates.containsKey(progressKey);
    }

    /**
     * すべてのタスクが完了するまで待機し、その後Runnableを実行します。
     * 保留中のプロセスがない場合も同様に実行します。
     * Runnableは最後にタスク数を更新したスレッドから実行されるため、長時間の処理は避けてください。
     * @param runnable タスクがなくなった時に実行するRunnable
     */
    public static void waitUntilDone(final Runnable runnable) {
        if(getTaskCount() == 0) {
            runnable.run();
            return;
        }
        TaskCountListener listener = new TaskCountListener() {
            @Override
            public void onUpdateTaskCount(int taskCount) {
                if(taskCount == 0) {
                    runnable.run();
                    removeTaskCountListener(this);
                }
            }
        };
        addTaskCountListener(listener);
    }

    /**
     * @return 現在のタスク数
     */
    public static synchronized int getTaskCount() {
        return sProgressStates.size();
    }

    /**
     * @return 実行中のタスクがあるかどうか
     */
    public static boolean hasOngoingTasks() {
        return getTaskCount() > 0;
    }

    /**
     * 進行状況レコードを保持する内部クラス。
     */
    public static class ProgressRecord {
        public final String key;
        public final int progress;
        public final int resid;
        public final Object[] varArg;

        public ProgressRecord(String key, int progress, int resid, Object[] varArg) {
            this.key = key;
            this.progress = progress;
            this.resid = resid;
            this.varArg = varArg;
        }
    }

    /**
     * @return 現在のすべての進行状況レコードのリスト
     */
    public static synchronized List<ProgressRecord> getProgressRecords() {
        if (sProgressStates.isEmpty()) return Collections.emptyList();
        List<ProgressRecord> records = new ArrayList<>();
        for (Map.Entry<String, ProgressState> entry : sProgressStates.entrySet()) {
            ProgressState state = entry.getValue();
            records.add(
                    new ProgressRecord(
                            entry.getKey(),
                            state.progress,
                            state.resid,
                            state.varArg == null ? new Object[0] : state.varArg.clone()
                    )
            );
        }
        return records;
    }
}

File: YukariLauncher/src/main/java/com/arata/yukarilauncher/task/ProgressKeeper.kt
package com.arata.yukarilauncher.task

import com.arata.yukarilauncher.task.TaskCountListener
import java.util.*

/** 進捗状態を管理するシングルトンオブジェクト。 */
object ProgressKeeper {

    /** 進捗リスナーマップ */
    private val sProgressListeners = HashMap<String, MutableList<ProgressListener>>()
    /** 進捗状態マップ */
    private val sProgressStates = HashMap<String, ProgressState>()
    /** タスクカウントリスナーリスト */
    private val sTaskCountListeners = ArrayList<TaskCountListener>()

    /** 進捗を登録し、リスナーに通知する。 @param progressRecord 進捗レコードキー @param progress 進捗値 @param resid リソースID @param va 追加引数 */
    @JvmStatic
    @Synchronized
    fun submitProgress(progressRecord: String, progress: Int, resid: Int, vararg va: Any?) {
        val progressState = sProgressStates[progressRecord]
        val shouldCallStarted = progressState == null
        val shouldCallEnded = resid == -1 && progress == -1
        if (shouldCallEnded) {
            sProgressStates.remove(progressRecord)
            updateTaskCount()
        } else if (shouldCallStarted) {
            val newState = ProgressState()
            sProgressStates[progressRecord] = newState
            updateTaskCount()
        }
        val currentState = sProgressStates[progressRecord]
        if (currentState != null) {
            currentState.progress = progress
            currentState.resid = resid
            currentState.varArg = va
        }

        val progressListeners = sProgressListeners[progressRecord]
        progressListeners?.forEach { listener ->
            when {
                shouldCallStarted -> listener.onProgressStarted()
                shouldCallEnded -> listener.onProgressEnded()
                else -> listener.onProgressUpdated(progress, resid, *va)
            }
        }
    }

    private fun notifyProgressUpdated(listener: ProgressListener, state: ProgressState) {
        listener.onProgressStarted()
        listener.onProgressUpdated(state.progress, state.resid, *(state.varArg ?: emptyArray<Any>()))
    }

    /** タスクカウントを更新し、リスナーに通知する。 */
    @Synchronized
    private fun updateTaskCount() {
        val count = sProgressStates.size
        for (listener in sTaskCountListeners) {
            listener.onUpdateTaskCount(count)
        }
    }

    /** 進捗リスナーを追加する。 @param progressRecord 進捗レコードキー @param listener リスナー */
    @JvmStatic
    @Synchronized
    fun addListener(progressRecord: String, listener: ProgressListener) {
        val state = sProgressStates[progressRecord]
        if (state != null && (state.resid != -1 || state.progress != -1)) {
            notifyProgressUpdated(listener, state)
        } else {
            listener.onProgressEnded()
        }
        val listenerList = sProgressListeners.computeIfAbsent(progressRecord) { ArrayList() }
        listenerList.add(listener)
    }

    /** 進捗リスナーを削除する。 @param progressRecord 進捗レコードキー @param listener リスナー */
    @JvmStatic
    @Synchronized
    fun removeListener(progressRecord: String, listener: ProgressListener) {
        sProgressListeners[progressRecord]?.remove(listener)
    }

    /** タスクカウントリスナーを追加する。 @param listener リスナー */
    @JvmStatic
    @Synchronized
    fun addTaskCountListener(listener: TaskCountListener) {
        listener.onUpdateTaskCount(sProgressStates.size)
        if (!sTaskCountListeners.contains(listener)) sTaskCountListeners.add(listener)
    }

    /** タスクカウントリスナーを追加する。 @param listener リスナー @param runUpdate 追加時に即時更新を実行するか */
    @JvmStatic
    @Synchronized
    fun addTaskCountListener(listener: TaskCountListener, runUpdate: Boolean) {
        if (runUpdate) listener.onUpdateTaskCount(sProgressStates.size)
        if (!sTaskCountListeners.contains(listener)) sTaskCountListeners.add(listener)
    }

    /** タスクカウントリスナーを削除する。 @param listener リスナー */
    @JvmStatic
    @Synchronized
    fun removeTaskCountListener(listener: TaskCountListener) {
        sTaskCountListeners.remove(listener)
    }

    /** 指定した進捗キーの進捗が存在するかを返す。 @param progressKey 進捗キー @return 存在するか */
    @JvmStatic
    fun containsProgress(progressKey: String): Boolean = sProgressStates.containsKey(progressKey)

    /** 全てのタスクが完了するまで待機してからRunnableを実行する。 @param runnable 実行する処理 */
    @JvmStatic
    fun waitUntilDone(runnable: Runnable) {
        if (taskCount == 0) {
            runnable.run()
            return
        }
        val listener = object : TaskCountListener {
            override fun onUpdateTaskCount(taskCount: Int) {
                if (taskCount == 0) {
                    runnable.run()
                    removeTaskCountListener(this)
                }
            }
        }
        addTaskCountListener(listener)
    }

    /** 現在のタスク数を返す。 */
    @JvmStatic
    @get:Synchronized
    val taskCount: Int
        get() = sProgressStates.size

    /** 進行中のタスクが存在するかを返す。 @return 進行中タスクが存在するか */
    @JvmStatic
    fun hasOngoingTasks(): Boolean = taskCount > 0

    /** 進捗レコード。 */
    class ProgressRecord(
        @JvmField val key: String,
        @JvmField val progress: Int,
        val resid: Int,
        val varArg: Array<Any?>
    )

    /** 全ての進捗レコードを取得する。 @return 進捗レコードのリスト */
    @JvmStatic
    @Synchronized
    fun getProgressRecords(): List<ProgressRecord> {
        if (sProgressStates.isEmpty()) return emptyList()
        val records = ArrayList<ProgressRecord>()
        for ((key, state) in sProgressStates) {
            records.add(
                ProgressRecord(
                    key,
                    state.progress,
                    state.resid,
                    state.varArg?.let { arg -> Array<Any?>(arg.size) { arg[it] } } ?: emptyArray<Any?>()
                )
            )
        }
        return records
    }
}
